// @arco-design/arco-scripts/lib/constant.js
"use strict";
var _a;
Object.defineProperty(exports, "__esModule", { value: true });
exports.ARCO_LIBRARY_PACKAGE_NAME_REACT = exports.FILENAME_README = exports.FILENAME_STYLE_ENTRY_CSS = exports.FILENAME_STYLE_ENTRY_RAW = exports.FILENAME_DIST_CSS = exports.FILENAME_DIST_LESS = exports.DIR_NAME_COMPONENT_LIBRARY = exports.DIR_NAME_DEMO = exports.DIR_NAME_ASSET = exports.DIR_NAME_SOURCE = exports.DIR_NAME_TEST = exports.DIR_NAME_SITE = exports.DIR_NAME_ICON = exports.DIR_NAME_UMD = exports.DIR_NAME_CJS = exports.DIR_NAME_ESM = exports.CWD = exports.BUILD_ENV_DIST_FILENAME_CSS = exports.BUILD_ENV_DIST_FILENAME_JS = exports.BUILD_ENV_TS_COMPILER = exports.BUILD_ENV_MODE = void 0;
_a = process.env, exports.BUILD_ENV_MODE = _a.BUILD_ENV_MODE, exports.BUILD_ENV_TS_COMPILER = _a.BUILD_ENV_TS_COMPILER, exports.BUILD_ENV_DIST_FILENAME_JS = _a.BUILD_ENV_DIST_FILENAME_JS, exports.BUILD_ENV_DIST_FILENAME_CSS = _a.BUILD_ENV_DIST_FILENAME_CSS;
exports.CWD = process.cwd();
exports.DIR_NAME_ESM = 'es';
exports.DIR_NAME_CJS = 'lib';
exports.DIR_NAME_UMD = 'dist';
exports.DIR_NAME_ICON = 'icon';
exports.DIR_NAME_SITE = 'site';
exports.DIR_NAME_TEST = 'tests';
exports.DIR_NAME_SOURCE = 'src';
exports.DIR_NAME_ASSET = 'asset';
exports.DIR_NAME_DEMO = 'demo';
exports.DIR_NAME_COMPONENT_LIBRARY = 'components';
exports.FILENAME_DIST_LESS = 'index.less';
exports.FILENAME_DIST_CSS = 'index.css';
exports.FILENAME_STYLE_ENTRY_RAW = 'index.js';
exports.FILENAME_STYLE_ENTRY_CSS = 'css.js';
exports.FILENAME_README = 'README.md';
exports.ARCO_LIBRARY_PACKAGE_NAME_REACT = '@arco-design/web-react';
